packages/core/src/utilities/VoxelManager.ts

```diff
@@ -308,6 +354,78 @@ export default class VoxelManager<T> {
     voxelManager.sourceVoxelManager = sourceVoxelManager;
     return voxelManager;
   }
+
+  /**
+   * Creates a history remembering voxel manager.
+   * This will remember the original values in the voxels, and will apply the
+   * update to the underlying source voxel manager.
+   */
+  public static createLazyVoxelManager<T>(
+    dimensions: Point3,
+    planeFactory: (width: number, height: number) => T
+  ): VoxelManager<T> {
+    const map = new Map<number, T>();
+    const [width, height, depth] = dimensions;
+    const planeSize = width * height;
+
+    const voxelManager = new VoxelManager(
+      dimensions,
+      (index) => map.get(Math.floor(index / planeSize))?.[index % planeSize],
+      (index, v) => {
+        const k = Math.floor(index / planeSize);
+        let layer = map.get(k);
+        if (!layer) {
+          layer = planeFactory(width, height);
+          map.set(k, layer);
+        }
+        layer[index % planeSize] = v;
+      }
+    );
+    voxelManager.map = map;
+    return voxelManager;
+  }
+
+  /**
+   * Creates a RLE based voxel manager.  This is effective for storing
+   * segmentation maps.
+   */
+  public static createRLEVoxelManager<T>(dimensions: Point3): VoxelManager<T> {
+    const [width, height, depth] = dimensions;
+    const map = new RLEVoxelMap<T>(width, height, depth);
+
+    const voxelManager = new VoxelManager<T>(
+      dimensions,
+      (index) => map.get(index),
+      (index, v) => map.set(index, v)
+    );
+    voxelManager.map = map;
+    voxelManager.getPixelData = map.getPixelData.bind(map);
+    return voxelManager;
+  }
+
+  /**
+   * This method adds a voxelManager instance to the image or volume object
+   * where the object added is of type:
+   * 1. RLE map if the scalar data is missing
+   * 2. Volume map for the imageId scalarData
+   */
+  public static addInstanceToImage(image: IImage) {
+    const { width, height } = image;
+    const scalarData = image.getPixelData();
+    if (scalarData?.length >= width * height) {
```

Comment:
this would fail if the image is stack (not volume ) and color right?

Reply:
It works for stack images, or the stack representation of volume images, both color (rgb or rgba), indexed and grayscale images.